[WIP] Special Relativistic Hydrodynamics

jzuhone opened the pull request:
This is not ready for review and merge quite yet--I need to make sure the rebase worked correctly first. 

Original checklist from @zengbs: 

- [ ] VL integrator
- [ ] PLM data reconstruction
- [ ] PPM data reconstruction
- [ ] HLLC Riemann solver
- [ ] HLLE Riemann solver
- [ ] AMR
- [ ] HDF5 output
- [ ] OpenMP
- [ ] GPU
- [ ] MPI
- [ ] Bitwise reproducibility
- [ ] yt analysis
- [ ] Record the SHA1 of commit in Record__Note
- [ ] Wiki [future work]

jzuhone commented:
@hyschive I am trying to check a segmentation fault with `-O0` but when I try this I get linking errors:

```
Object/__cpu__Flu_Close.o: In function `CorrectUnphysical(int, int, int const*, double const (*) [5][10648], double (*) [5][4096], char (*) [4096], double (*) [9][5][256], double const (*) [0][11132], double const (*) [0][4352], double) [clone ._omp_fn.0]':
/home6/jzuhone/gamer/src/Fluid/Flu_Close.cpp:915: undefined reference to `Hydro_CheckMinEintInEngy(double, double, double, double, double, double, double)'
/home6/jzuhone/gamer/src/Fluid/Flu_Close.cpp:793: undefined reference to `Hydro_RiemannSolver_Roe(int, double*, double const*, double const*, double, double, double (*)(double, double, double const*, double const*, int const*, double const* const*), double (*)(double, double, double const*, double const*, int const*, double const* const*), double const*, int const*, double const* const*)'
/home6/jzuhone/gamer/src/Fluid/Flu_Close.cpp:796: undefined reference to `Hydro_RiemannSolver_Roe(int, double*, double const*, double const*, double, double, double (*)(double, double, double const*, double const*, int const*, double const* const*), double (*)(double, double, double const*, double const*, int const*, double const* const*), double const*, int const*, double const* const*)'
/home6/jzuhone/gamer/src/Fluid/Flu_Close.cpp:647: undefined reference to `Hydro_RiemannSolver_Roe(int, double*, double const*, double const*, double, double, double (*)(double, double, double const*, double const*, int const*, double const* const*), double (*)(double, double, double const*, double const*, int const*, double const* const*), double const*, int const*, double const* const*)'
/home6/jzuhone/gamer/src/Fluid/Flu_Close.cpp:654: undefined reference to `Hydro_RiemannSolver_Roe(int, double*, double const*, double const*, double, double, double (*)(double, double, double const*, double const*, int const*, double const* const*), double (*)(double, double, double const*, double const*, int const*, double const* const*), double const*, int const*, double const* const*)'
collect2: error: ld returned 1 exit status
```

I've tried to resolve them but cannot figure them out.

hyschive commented:
@jzuhone I guess it's caused by this [line](https://github.com/jzuhone/gamer/blob/srhd/src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp#L683) (other linking problems are also caused by similar conditional compilation problems). What I don't understand is why the code can be compiled with `-O3`.

hyschive commented on src/Model_Hydro/CPU_Hydro/CPU_dtSolver_HydroCFL.cpp:
```diff
@@ -99,19 +108,32 @@ void CPU_dtSolver_HydroCFL  ( real g_dt_Array[], const real g_Flu_Array[][FLU_NI
+
+#        ifdef SRHD
+         real Rho;
+         Hydro_Con2Pri( fluid, Pri,(real)NULL_REAL, NULL_BOOL, NULL_INT, NULL, NULL_BOOL,
+                        (real)NULL_REAL, NULL, NULL,
+                        EoS.GuessHTilde_FuncPtr, EoS.HTilde2Temp_FuncPtr,
+                        EoS.AuxArrayDevPtr_Flt, EoS.AuxArrayDevPtr_Int, EoS.Table, NULL, &LorentzFactor );
```
- The input parameters are inconsistent with `Hydro_Con2Pri()`.
- Change `FLU_NIN_T` from `NCOMP_FLUID` to `NCOMP_TOTAL` in `Macro.h` for SRHD if it's really necessary to call `Hydro_Con2Pri()` here.

jzuhone replied:
Hi @hyschive, thanks--I realize that this has fallen behind `master`, unfortunately I have been tied up with a telescope proposal. Hoping to get back to this within a few weeks.

hyschive replied:
@jzuhone Thanks! I just accidentally found this bug. Feel free to let me know if there is anything I can help.

jzuhone replied:
@hyschive does https://github.com/gamer-project/gamer/pull/185/commits/7ef64604e13f614cc26dffe6c2a7a7f4bb9a1385 address it?

hyschive replied:
@jzuhone I just realized that I was checking a different version of the SRHD code. So some of my comments are actually incorrect. Sorry for the false alarm!

I'll add comments on those commits directly.

jzuhone commented:
@hyschive separate issue. I tried the setup with the jet incoming at the bottom of the domain that we used in the Galaxies paper. This simulation was run with the @zengbs branch. Currently I am getting errors like these (added some other messages for information):

```
WARNING : parameter [DT__FLUID                   ] is reset to [ 4.00000000000000e-01]
WARNING : parameter [DT__FLUID_INIT              ] is reset to [ 4.00000000000000e-01]
WARNING : parameter [OPT__FLU_INT_SCHEME         ] is reset to [ 4                   ]
WARNING : parameter [OPT__REF_FLU_INT_SCHEME     ] is reset to [ 4                   ]
WARNING : parameter [OPT__1ST_FLUX_CORR          ] is reset to [ 2                   ] for HYDRO
WARNING : parameter [OPT__TIMING_BARRIER         ] is reset to [ 0                   ]
WARNING : parameter [MIN_PRES                    ] is reset to [ 6.66666666666667e-06]
WARNING : parameter [OPT__CHECK_PRES_AFTER_FLU   ] is reset to [ 0                   ]
WARNING : parameter [MU_NORM                     ] is reset to [ 1.67353275798800e-24]
WARNING : parameter [FLAG_BUFFER_SIZE_MAXM2_LV   ] is reset to [ 6                   ]
WARNING : parameter [OPT__RESET_FLUID_INIT       ] is reset to [ 0                   ] to match OPT__RESET_FLUID
WARNING : please make sure that different MPI ranks will use different GPUs for "OPT__GPUID_SELECT == -1" !!
WARNING : it's recommended to use the outflow BC (currently OPT__BC_FLU[2] = 4 != 2)
REMINDER : OPT__1ST_FLUX_CORR may break the strict conservation of fluid variables
WARNING : MIN_DENS (1.0000000e-05) is on --> please ensure that this value is reasonable !!
WARNING : MIN_PRES (6.6666667e-06) is on --> please ensure that this value is reasonable !!
WARNING : MIN_EINT (1.0000000e-05) is on --> please ensure that this value is reasonable !!
WARNING : MIN_TEMP == 0.0 could be dangerous and is mainly for debugging only !!
WARNING : MIN_ENTR == 0.0 could be dangerous and is mainly for debugging only !!
ERROR : Interpolate_Iterate() failed !!
NComp=5, IntScheme=4, UnwrapPhase=0, Monotonic=1, OppSign0thOrder=1
FData_is_Prim=0, Iter=0, IntMonoCoeff=2.0000000e+00
Fluid:  [0]= 8.5108856e-03 [1]=-3.1494386e-04 [2]= 9.5832104e-02 [3]= 3.4397380e-03 [4]= 6.6977082e-02
``` 

I note that `Interpolate_Iterate` was not in GAMER in the previous SRHD code, so there may be something that I am missing here.

hyschive commented:
> @hyschive separate issue. I tried the setup with the jet incoming at the bottom of the domain that we used in the Galaxies paper. This simulation was run with the @zengbs branch. Currently I am getting errors like these (added some other messages for information):
> 
> ```
> WARNING : parameter [DT__FLUID                   ] is reset to [ 4.00000000000000e-01]
> WARNING : parameter [DT__FLUID_INIT              ] is reset to [ 4.00000000000000e-01]
> WARNING : parameter [OPT__FLU_INT_SCHEME         ] is reset to [ 4                   ]
> WARNING : parameter [OPT__REF_FLU_INT_SCHEME     ] is reset to [ 4                   ]
> WARNING : parameter [OPT__1ST_FLUX_CORR          ] is reset to [ 2                   ] for HYDRO
> WARNING : parameter [OPT__TIMING_BARRIER         ] is reset to [ 0                   ]
> WARNING : parameter [MIN_PRES                    ] is reset to [ 6.66666666666667e-06]
> WARNING : parameter [OPT__CHECK_PRES_AFTER_FLU   ] is reset to [ 0                   ]
> WARNING : parameter [MU_NORM                     ] is reset to [ 1.67353275798800e-24]
> WARNING : parameter [FLAG_BUFFER_SIZE_MAXM2_LV   ] is reset to [ 6                   ]
> WARNING : parameter [OPT__RESET_FLUID_INIT       ] is reset to [ 0                   ] to match OPT__RESET_FLUID
> WARNING : please make sure that different MPI ranks will use different GPUs for "OPT__GPUID_SELECT == -1" !!
> WARNING : it's recommended to use the outflow BC (currently OPT__BC_FLU[2] = 4 != 2)
> REMINDER : OPT__1ST_FLUX_CORR may break the strict conservation of fluid variables
> WARNING : MIN_DENS (1.0000000e-05) is on --> please ensure that this value is reasonable !!
> WARNING : MIN_PRES (6.6666667e-06) is on --> please ensure that this value is reasonable !!
> WARNING : MIN_EINT (1.0000000e-05) is on --> please ensure that this value is reasonable !!
> WARNING : MIN_TEMP == 0.0 could be dangerous and is mainly for debugging only !!
> WARNING : MIN_ENTR == 0.0 could be dangerous and is mainly for debugging only !!
> ERROR : Interpolate_Iterate() failed !!
> NComp=5, IntScheme=4, UnwrapPhase=0, Monotonic=1, OppSign0thOrder=1
> FData_is_Prim=0, Iter=0, IntMonoCoeff=2.0000000e+00
> Fluid:  [0]= 8.5108856e-03 [1]=-3.1494386e-04 [2]= 9.5832104e-02 [3]= 3.4397380e-03 [4]= 6.6977082e-02
> ```
> 
> I note that `Interpolate_Iterate` was not in GAMER in the previous SRHD code, so there may be something that I am missing here.

@jzuhone Could you share both the original code used in the Galaxies paper that works well and the new code from this PR that triggers the above error message? I'll try to reproduce it locally first.

jzuhone commented:
@hyschive 

the branch that worked for me last year is here:

https://github.com/jzuhone/gamer/tree/jet_icm_wall

and the `Input__*` files and `Makefile` are here: 

```
curl -JO http://use.yt/upload/927afd1e
```

jzuhone commented:
@hyschive and sorry, in the files I just sent, I would change `MAX_LEVEL = 3` in `Input__Parameter`.

hyschive commented:
> @hyschive and sorry, in the files I just sent, I would change `MAX_LEVEL = 3` in `Input__Parameter`.

@jzuhone
### Short answer
Adopting `OPT__FLU_INT_SCHEME = OPT__REF_FLU_INT_SCHEME = 4` (or `-1` by default) and `MONO_MAX_ITER > 0` (or `-1` by default) solves the problem.

### Long answer
First of all, when interpolation causes unphysical results (e.g., negative pressure), GAMER adopts two approaches to try to solve the problem: using primitive instead of conservation variables for interpolation ([IntPrim](https://github.com/jzuhone/gamer/blob/srhd/src/Interpolation/Interpolate.cpp#L76-L77)) or reducing the monotonic coefficient to make interpolation more diffusive ([ReduceMonoCoeff](https://github.com/jzuhone/gamer/blob/srhd/src/Interpolation/Interpolate.cpp#L78-L79)). The disadvantage of `IntPrim` is that it breaks conservation.

The new version (i.e., the `srhd` branch) always disables `IntPrim` and only uses `ReduceMonoCoeff` during grid refinement ([here](https://github.com/jzuhone/gamer/blob/srhd/src/LoadBalance/LB_Refine_AllocateNewPatch.cpp#L970-L971)) to ensure conservation. However, `OPT__FLU_INT_SCHEME = 3` does not support `ReduceMonoCoeff` (so one must set `MONO_MAX_ITER=0`), and thus the code will just be terminated when detecting unphysical results during grid refinement.

On the contrary, the old version (i.e., the `jet_icm_wall` branch) always enables `IntPrim` (see [here](https://github.com/jzuhone/gamer/blob/jet_icm_wall/src/Interpolation/AdaptiveInterpolate.cpp#L31-L46)) during grid refinement, which eliminates unphysical results but sacrifices conservation.

### Other minor comments
- `SRHD_CheckUnphysical()` in the old version also checks `HTilde` and `Temp` while `Hydro_CheckUnphysical()` in the new code doesn't. I'm not sure which one is more appropriate.
- There are lots of warnings about negative passive scalars during flux fix-up (triggered [here](https://github.com/jzuhone/gamer/blob/srhd/src/Fluid/Flu_FixUp_Flux.cpp#L231) and then [here](https://github.com/jzuhone/gamer/blob/srhd/src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp#L834-L839)). An example is given below. I don't know whether they would affect your results. I guess the same issue also exists in the old version but is not shown since `SRHD_CheckUnphysical()` doesn't check passive scalars.
```
ERROR: unphysical conserved variables at file <Fluid/Flu_FixUp_Flux.cpp>, line <231>, function <Flu_FixUp_Flux
D= 1.0453720e+03, Mx= 3.4694470e-19, My=-4.0711792e-18, Mz= 0.0000000e+00, E= 2.6509909e-02
E^2+2*E*D-|M|^2= 5.0719257e-05
Passive: [0]=2.2250739e-308 [1]=1.0453720e+03 [2]=-5.2099325e-26 [3]=2.2250739e-308
```
- Consider adding `example/test_problem/Hydro/Jets/JetICMWall`, including analysis and visualization scripts.

@zengbs Feel free to share any comments.

hyschive commented:
BTW, in the new version one needs to replace `DT_SPEED_OF_LIGHT` with `DT__SPEED_OF_LIGHT` in `Input__Parameter`.

jzuhone commented:
@hyschive please check what I have done in https://github.com/gamer-project/gamer/pull/185/commits/09027ae7c8af0d47d027bb302111d58f36d1f4a4, I am not certain about it.

jzuhone commented:
@hyschive any thoughts about https://github.com/gamer-project/gamer/commit/09027ae7c8af0d47d027bb302111d58f36d1f4a4? The issue is how I handled the passive scalars.

hyschive commented:
@jzuhone I'm starting to review this PR. One quick question. I just found that the test problem `Riemann` supports user-specified left and right states in the original [PR](https://github.com/gamer-project/gamer/pull/41/files#diff-a8d622edec7ff4c03efddb40e9646cd77d7683cd844c9090ee4fd273b9441ddd) by @zengbs but not in this new PR. I haven't checked other files or test problems so I'm not sure if there are missing updates in other places too. Should I start reviewing it or should I wait for the green light from you?

jzuhone commented:
@hyschive please go ahead and start reviewing.

ChunYen-Chen commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+  if ( *Constant >= Discrimination )
+  {
+     // Eq. A6 in "Tseng et al. 2021, MNRAS, 504, 3298"
+     GuessHTilde = (real)1.3333333 * SQRT( *Constant );
```
Should we change the number `1.3333333` to `(real)4./3.` since the number of digits is insufficient for double precision?

zengbs replied:
Btw, I have confirmed that using this approximate value as an initial guess will still give accurate results (see Fig. 2 in Tseng 2021) after multiple iterations.

hyschive replied:
For clarity, let's use `(real)4./3.`. It won't hurt performance since it's just a literal determined during compilation.

ChunYen-Chen commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+  else
+  {
+     // Eq. A4 in "Tseng et al. 2021, MNRAS, 504, 3298"
+     GuessHTilde  = (real)11.18 * SQRT( B*C* (*Constant) + (real)45.0*B*B );
```
Should we change the number `11.18` to `SQRT((real)125.) since the number of digits is insufficient for double precision?

zengbs replied:
GuessHTilde is just an initial guess value before starting root finding. Using 11.18, rather than sqrt(125.0), can slightly improve computing speed.

ChunYen-Chen replied:
Thanks for the explanation!

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_dtSolver_HydroCFL.cpp:
```diff
@@ -99,19 +108,32 @@ void CPU_dtSolver_HydroCFL  ( real g_dt_Array[], const real g_Flu_Array[][FLU_NI
-#        else
+#        elif ( !defined SRHD )
          Emag = NULL_REAL;
 #        endif
```
`defined MHD == false && defined SRHD` case is not included.

ChunYen-Chen commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+// Note        :  Eq. 16 in "Tseng et al. 2021, MNRAS, 504, 3298"
+//
+// Parameter   :  HTilde     : Reduced enthalpy
+//                Temp       : Temperature (kT/mc**)
```
`mc**` -> `mc**2`

ChunYen-Chen commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+// Function    :  EoS_HTilde2Temp_TaubMathews
+// Description :  Convert reduced enthalpy to temperature
+//
+// Note        :  Eq. 16 in "Tseng et al. 2021, MNRAS, 504, 3298"
```
Also Eq A.4 for the `DiffTemp`

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_dtSolver_HydroCFL.cpp:
```diff
@@ -147,6 +180,7 @@ void CPU_dtSolver_HydroCFL  ( real g_dt_Array[], const real g_Flu_Array[][FLU_NI
 //       no longer used
          MaxCFL = FMAX( CFLx+CFLy+CFLz, MaxCFL );
          */
+#        endif
```
How about moving line `161` which is `MaxCFL = FMAX( Us_Max * LorentzFactor_Max + LorentzFactor_s_Max * U_Max, MaxCFL );` to here for clarity?

ChunYen-Chen commented on src/Output/Output_L1Error.cpp:
```diff
@@ -367,15 +372,10 @@ void WriteFile( void (*AnalFunc_Flu)( real fluid[], const double x, const double
+   Anal[ENGY] = Hydro_Con2Pres( Anal[DENS], Anal[MOMX], Anal[MOMY], Anal[MOMZ], Anal[ENGY], Anal+NCOMP_FLUID,
+                                CheckMinPres_No, NULL_REAL, Emag_Zero,
+                                EoS_DensEint2Pres_CPUPtr, EoS_GuessHTilde_CPUPtr, EoS_HTilde2Temp_CPUPtr,
+                                EoS_AuxArray_Flt, EoS_AuxArray_Int, h_EoS_Table, NULL );
```
The analytical solution of temperature is missing here.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_CTU.cpp:
Please update the comment at the end of the file to this: 
`#endif // #if ( MODEL == HYDRO  &&  FLU_SCHEME == CTU  &&  !defined SRHD )`

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_CTU.cpp:
```diff
@@ -1,6 +1,6 @@
 #include "CUFLU.h"
 
-#if ( MODEL == HYDRO  &&  FLU_SCHEME == CTU )
+#if ( MODEL == HYDRO  &&  FLU_SCHEME == CTU && !defined SRHD )
```
Please update the code style to two spaces between `&&` and the conditions.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_RTVD.cpp:
Please update the comment at the end of the file to this:
#endif // #if ( MODEL == HYDRO  &&  FLU_SCHEME == RTVD  &&  !defined SRHD )

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_RTVD.cpp:
```diff
@@ -1,6 +1,6 @@
 #include "CUFLU.h"
 
-#if ( !defined GPU  &&  MODEL == HYDRO  &&  FLU_SCHEME == RTVD )
+#if ( !defined GPU  &&  MODEL == HYDRO  &&  FLU_SCHEME == RTVD && !defined SRHD )
```
Please update the code style to two spaces between && and the conditions.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:
```diff
@@ -1993,27 +2002,33 @@ void Hydro_HancockPredict( real fc[][NCOMP_LR], const real dt, const real dh,
          break;  // no need to apply the floors since the input values should already satisfy these constraints
       }
 
+#ifndef SRHD
```
Please add spaces to align the current indent.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:
```diff
@@ -1993,27 +2002,33 @@ void Hydro_HancockPredict( real fc[][NCOMP_LR], const real dt, const real dh,
+      for (int v=NCOMP_FLUID; v<NCOMP_TOTAL; v++) fcCon[f][v] = FMAX( fcCon[f][v], TINY_NUMBER );
 #     endif
    } // for (int f=0; f<6; f++)
 #  endif // #ifdef MHM_CHECK_PREDICT
+#endif // #ifndef SRHD
```
The order of the comments is wrong. `\\ #ifdef MHM_CHECK_PREDICT` is before `// #ifndef SRHD`. Also, please add spaces to align the current indent.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:
```diff
@@ -1993,27 +2002,33 @@ void Hydro_HancockPredict( real fc[][NCOMP_LR], const real dt, const real dh,
       if ( fc[f][4] <= (real)0.0 || fc[f][4] >= HUGE_NUMBER || fc[f][4] != fc[f][4] ) reset_cell = true;
```
Please change all the `fc` to `fcCon`.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:
```diff
@@ -1993,27 +2002,33 @@ void Hydro_HancockPredict( real fc[][NCOMP_LR], const real dt, const real dh,
       if ( fc[f][4] <= (real)0.0 || fc[f][4] >= HUGE_NUMBER || fc[f][4] != fc[f][4] ) reset_cell = true;
       if ( Pres     <= (real)0.0 || Pres     >= HUGE_NUMBER || Pres     != Pres     ) reset_cell = true;
```
If we passed the primitive variables into `Hydro_HancockPredict`, we should not calculate the `Pres` in the function once more at line 1999.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:
```diff
@@ -1926,7 +1927,7 @@ void Hydro_LimitSlope( const real L[], const real C[], const real R[], const LR_
 //                3. Input variables must be conserved variables
 //
-// Parameter   :  fc                : Face-centered conserved variables to be updated
+// Parameter   :  fcCon             : Face-centered conserved variables to be updated
```
Please also add the description about `fcPri`. Also, we should notice the user `fcPri` will not be updated.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DualEnergy.cpp:
Please update the comment at the end of the file to this:
`#endif // #if ( MODEL == HYDRO  &&  defined DUAL_ENERGY  &&  !defined SRHD )`

ChunYen-Chen commented on example/test_problem/Hydro/JetICMWall/Input__Flag_RhoGradient:
The second column is not aligned.

ChunYen-Chen commented on src/Fluid/Flu_FixUp_Restrict.cpp:
```diff
@@ -297,7 +297,7 @@ void Flu_FixUp_Restrict( const int FaLv, const int SonFluSg, const int FaFluSg,
 //    check the minimum pressure/internal energy and, when the dual-energy formalism is adopted, ensure the consistency between
 //    pressure, total energy density, and the dual-energy variable
-#     if ( MODEL == HYDRO )
+#     if ( MODEL == HYDRO && !defined SRHD )
```
Please also update the comment at line 347 to be `# endif // #if ( MODEL == HYDRO  &&  !defined SRHD )`, and there should be two spaces around `&&` to match the code style.

ChunYen-Chen commented on src/Fluid/Flu_ResetByUser.cpp:
```diff
@@ -291,7 +291,7 @@ void Flu_ResetByUser_API_Default( const int lv, const int FluSg, const int MagSg
          if ( Reset )
          {
-#           if ( MODEL == HYDRO )
+#           if ( MODEL == HYDRO && !defined SRHD )
```
Please also update the comment at line 313 to be `# endif // #if ( MODEL == HYDRO  &&  !defined SRHD )`, and there should be two spaces around && to match the code style.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_RiemannSolver_Exact.cpp:
```diff
@@ -5,7 +5,7 @@
 #include "CUFLU.h"
 
-#if (  MODEL == HYDRO  &&  \
+#if (  MODEL == HYDRO  &&  !defined SRHD  &&  \
```
Please also update the comment at line 425.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_RiemannSolver_Exact.cpp:
```diff
@@ -88,9 +89,9 @@ void Hydro_RiemannSolver_Exact( const int XYZ, real Flux_Out[], const real L_In[
 // convert conserved variables to primitive variables
    Hydro_Con2Pri( L_In, L, MinPres, FracPassive_No, NULL_INT, NULL, JeansMinPres_No, NULL_REAL,
-                  EoS_DensEint2Pres, NULL, EoS_AuxArray_Flt, EoS_AuxArray_Int, EoS_Table, NULL );
+                  EoS_DensEint2Pres, NULL, NULL, NULL, EoS_AuxArray_Flt, EoS_AuxArray_Int, EoS_Table, NULL, NULL );
```
The `EoS_GuessHTilde`, `EoS_HTilde2Temp`, and `LorentzFactorPtr` should be `NULL` if `SRHD` is not enabled. Should we replace the `NULL` to be the variables respectively? (Same question to the other `NULL`s)

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_RiemannSolver_HLLD.cpp:
```diff
@@ -5,7 +5,7 @@
 #include "CUFLU.h"
 
-#if ( MODEL == HYDRO  &&  defined MHD )
+#if ( MODEL == HYDRO  &&  defined MHD  &&  !defined SRHD )
```
Please also update the comment at line 448.

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_RiemannSolver_Roe.cpp:
```diff
@@ -5,7 +5,7 @@
 #include "CUFLU.h"
 
-#if ( MODEL == HYDRO )
+#if ( MODEL == HYDRO  &&  !defined SRHD )
```
Please also update the comment at line 712.

ChunYen-Chen commented on src/Model_Hydro/GPU_Hydro/CUFLU_FluidSolver_RTVD.cu:
Please change the line 4 to be `#if ( defined GPU  &&  MODEL == HYDRO  &&  FLU_SCHEME == RTVD  &&  !defined SRHD )` and also update the comment at line 420 accordingly.

ChunYen-Chen commented on src/Refine/Flag_Real.cpp:
```diff
@@ -296,11 +301,46 @@ void Flag_Real( const int lv, const UseLBFunc_t UseLBFunc )
+						Fluid[MOMZ][k][j][i], Fluid[ENGY][k][j][i] };
+		     real U1, U2, U3;
+
+#  ifdef CHECK_UNPHYSICAL_IN_FLUID
+		     Hydro_CheckUnphysical( UNPHY_MODE_CONS, Cons, NULL, ERROR_INFO, UNPHY_VERBOSE );
+#  endif
```
Please add spaces to match the indent.

ChunYen-Chen commented on src/Refine/Refine.cpp:
```diff
@@ -833,7 +833,7 @@ void Refine( const int lv, const UseLBFunc_t UseLBFunc )
 
 
-#           if ( MODEL == HYDRO )
+#           if ( MODEL == HYDRO && !defined SRHD )
```
Please also update the comment at line 866.

ChunYen-Chen commented on src/TestProblem/Hydro/BlastWave/Init_TestProb_Hydro_BlastWave.cpp:
```diff
@@ -223,6 +223,17 @@ void SetGridIC( real fluid[], const double x, const double y, const double z, co
+#  ifdef SRHD
+   real Prim[NCOMP_TOTAL];
+   Prim[0] = Dens;
+   Prim[1] = MomX;
+   Prim[2] = MomY;
+   Prim[3] = MomZ;
```
It is a little bit confusing to assign a primitive variable to a conserved variable. Can you change it to `Prim[*] = 0.0;`?

ChunYen-Chen commented on src/Auxiliary/Aux_Check_Parameter.cpp:
```diff
@@ -650,14 +655,21 @@ void Aux_Check_Parameter()
 #   if ( RSOLVER_RESCUE != NONE  &&  RSOLVER_RESCUE != ROE  &&  RSOLVER_RESCUE != HLLE  &&  RSOLVER_RESCUE != HLLD )
 #     error : ERROR : unsupported RSOLVER_RESCUE for MHD (ROE/HLLE/HLLD) !!
 #   endif
+#  elif SRHD 
+#   if ( defined RSOLVER  &&  RSOLVER != HLLC  &&  RSOLVER != HLLE )
+#     error : ERROR : unsupported Riemann solver for SRHD (HLLC/HLLE) !!
+#   endif
+#   if ( defined FLU_SCHEME  &&  FLU_SCHEME != MHM_RP && FLU_SCHEME != MHM )
+#     error : ERROR : unsupported FLU_SCHEME for SRHD (MHM_RP/MHM) !!
+#   endif
```
How about separating this section into an individual `#if` section? Also, we need to add more checks for `SRHD`,  e.g. `GRAVITY` should be `false`, `MHD` should be `false`, etc.

ChunYen-Chen commented on src/Fluid/Flu_BoundaryCondition_User.cpp:
```diff
@@ -36,8 +38,9 @@ extern void (*BC_BField_User_Ptr)( real magnetic[], const double x, const double
 //-------------------------------------------------------------------------------------------------------
-void BC_User_Template( real fluid[], const double x, const double y, const double z, const double Time,
-                       const int lv, double AuxArray[] )
+void BC_User_Template( real Array[], const int ArraySize[], real fluid[], const int NVar_Flu,
+		       const int GhostSize, const int idx[], const double pos[], const double Time,
+		       const int lv, const int TFluVarIdxList[], double AuxArray[] )
```
Please also update the `Parameter` section of the function description.

ChunYen-Chen commented on src/Fluid/Flu_BoundaryCondition_User.cpp:
```diff
@@ -112,10 +120,10 @@ void BC_User_Template( real fluid[], const double x, const double y, const doubl
-                                 const int ArraySizeZ, const int Idx_Start[], const int Idx_End[],
-                                 const int TFluVarIdxList[], const double Time, const double dh, const double *Corner,
-                                 const long TVar, const int lv )
+void Flu_BoundaryCondition_User( real *Array, const int NVar_Flu, const int GhostSize, const int ArraySizeX,
+                                 const int ArraySizeY, const int ArraySizeZ, const int Idx_Start[],
+                                 const int Idx_End[], const int TFluVarIdxList[], const double Time,
+                                 const double dh, const double *Corner, const long TVar, const int lv )
```
Please also update the `Parameter` section of the function description.

ChunYen-Chen commented on src/Fluid/Flu_Close.cpp:
```diff
@@ -88,7 +90,7 @@ void Flu_Close( const int lv, const int SaveSg_Flu, const int SaveSg_Mag,
 // try to correct the unphysical results in h_Flu_Array_F_Out (e.g., negative density)
 // --> must be done BEFORE invoking both StoreFlux() and CorrectFlux() since CorrectUnphysical() might modify the flux array
-#  if ( MODEL == HYDRO )
+#  if ( MODEL == HYDRO && !defined SRHD )
```
We should also update the function prototype accordingly (from line 17).

ChunYen-Chen commented on src/Fluid/Flu_Close.cpp:
```diff
@@ -447,6 +451,7 @@ bool Unphysical( const real Fluid[], const int CheckMode, const real Emag )
 } // FUNCTION : Unphysical
 
 
+#ifndef SRHD
```
We should also update the function prototype accordingly (from line 18).

ChunYen-Chen commented on src/Fluid/Flu_DerivedField_BuiltIn.cpp:
```diff
@@ -32,6 +33,11 @@ void Flu_DerivedField_DivVel( real Out[], const real FluIn[], const real MagIn[]
 {
 
+#  ifdef SRHD
+   Aux_Error( ERROR_INFO, "SRHD does not support div(V) evalation !!\n" );
```
A typo of `evaluation`?

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:
```diff
@@ -994,6 +1277,15 @@ real Hydro_Con2Entr( const real Dens, const real MomX, const real MomY, const re
 
+#  if (  defined SRHD  &&  defined  GAMER_DEBUG )
+#  ifdef __CUDACC__
+   printf( "ERROR : SRHD does not support entropy evalation at file <%s>, line <%d>, function <%s> !!\n",
```
A typo of `evaluation`?

ChunYen-Chen commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:
```diff
@@ -994,6 +1277,15 @@ real Hydro_Con2Entr( const real Dens, const real MomX, const real MomY, const re
+   printf( "ERROR : SRHD does not support entropy evalation at file <%s>, line <%d>, function <%s> !!\n",
+           ERROR_INFO );
+#  else
+   Aux_Error( ERROR_INFO, "SRHD does not support entropy evalation !!\n" );
```
A typo of `evaluation`?

ChunYen-Chen commented on src/Fluid/Flu_FixUp_Flux.cpp:
```diff
@@ -148,7 +148,7 @@ void Flu_FixUp_Flux( const int lv )
 
 //             calculate the internal energy density and pressure
-#              if ( MODEL == HYDRO  &&  !defined BAROTROPIC_EOS )
+#              if ( MODEL == HYDRO  &&  !defined BAROTROPIC_EOS && !defined SRHD )
```
Please also update the comment at linen 216

ChunYen-Chen commented on src/Fluid/Flu_FixUp_Flux.cpp:
```diff
@@ -267,7 +272,7 @@ void Flu_FixUp_Flux( const int lv )
 //                --> assuming the variable "Eint" is correct
 //                --> no need to check the internal energy floor here since we have skipped failing cells
-#                 if ( MODEL == HYDRO )
+#                 if ( MODEL == HYDRO && !defined SRHD )
```
Please also update the comment at line 294.

ChunYen-Chen commented on src/Init/Init_Load_Parameter.cpp:
```diff
@@ -504,7 +510,7 @@ void Init_Load_Parameter()
    ReadPara->Add( "OPT__CK_PATCH_ALLOCATE",     &OPT__CK_PATCH_ALLOCATE,          false,           Useless_bool,  Useless_bool   );
    ReadPara->Add( "OPT__CK_FLUX_ALLOCATE",      &OPT__CK_FLUX_ALLOCATE,           false,           Useless_bool,  Useless_bool   );
-#  if ( MODEL == HYDRO )
+#  if ( MODEL == HYDRO && defined SRHD )
```
`OPT__CK_NEGATIVE` is not assigned if `MODEL == HYDRO && !defined SRHD` which might cause some problems at `src/Auxiliary/Aux_Check.cpp` at line 39 and `src/Aux_TakeNote.cpp` at line 1461.

ChunYen-Chen commented:
@zengbs Hi, I noticed some inconsistencies between your codes (`EoS_GuessHTilde_TaubMathews()`) and your paper. The inconsistency should not change the simulation result since the function only provides an initial guess, but this kind of initial guess might hurt the performance when finding roots. 

The attached figure is my checks from the code and your paper. The `HTilde` you guessed in the code is different from both the equation (A4) of your paper and the solution of equation (A3). Please do let me know if any of my derivation is wrong.
![SRHD](https://github.com/gamer-project/gamer/assets/70311975/aba453c6-4f6a-4ce9-b224-f38ffcf4774e)

zengbs commented:
Can you use Sympy to check if two expressions are equivalent?

jzuhone commented:
@ChunYen-Chen you had a bunch of other comments here but now they are gone somehow?

ChunYen-Chen reviewed (changes requested):
Hi, @jzuhone. Sorry for the previous confusing comments. I have completed my initial review.  There are some significant bugs in the file listed below. These bugs are indicated with [BUG] at the beginning of the comments. The other comments relate to code optimization and minor issues. Please feel free to let me know if any of the comments are unclear or if you disagree with any suggestions.

### Bug files
- `src/Init/Init_Load_Parameter.cpp`
- `src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp`
- `src/Model_Hydro/GPU_Hydro/CUFLU_FluidSolver_RTVD.cu`
- `src/Output/Output_L1Error.cpp`

ChunYen-Chen reviewed (changes requested):
Hi @jzuhone, 
There are more bugs in the other files. Please also make sure those are fixed. Thanks!

jzuhone commented:
@zengbs I do not seem to have implemented the passive scalars correctly here. Do you have any thoughts on what is wrong?

ChunYen-Chen commented:
Hi @jzuhone,
Could you please explain what's happening with the passive scalar, and could you provide guidance on how to reproduce the error? I might be able to assist :)

ChunYen-Chen commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+     // Eq. A4 in "Tseng et al. 2021, MNRAS, 504, 3298"
+     GuessHTilde  = (real)11.18 * SQRT( B*C* (*Constant) + (real)45.0*B*B );
+	 GuessHTilde -= (real)75.0 * B;
+	 GuessHTilde /= C;
```
This part of the code has been identified as incorrect, but the EoS results should be correct, as this function only provides an initial guess.

jzuhone commented:
@ChunYen-Chen it happens in the `JetICMWall` test problem I added. The passive scalars are going negative. I attach a couple of pictures, one showing a slice of density (looks fine) and another showing a slice of one of the passive scalars (a mess). 

The passive scalars in this problem worked ok in @zengbs 's previous branch, which we used in this paper: https://arxiv.org/abs/2303.02296 (see Figure 5).

![Data_000030_Slice_z_density](https://github.com/gamer-project/gamer/assets/1914976/84f1561e-06f6-42e8-bc12-8cbc49cf7318)

![Data_000029_Slice_z_int](https://github.com/gamer-project/gamer/assets/1914976/b631957c-1b61-43fd-b02a-37c56c67a340)

ChunYen-Chen commented:
@jzuhone I see~ Is there anything that needs to be modified in the default `JetICMWall/Input__*` to reproduce the negative passive scalar?

jzuhone commented:
@ChunYen-Chen Nope, should be able to reproduce it with the normal setup

ChunYen-Chen commented:
Hi @jzuhone,
I noticed there are numerous error messages regarding unphysical conserved variables in the `log`:
```
ERROR: unphysical conserved variables at file <Fluid/Flu_FixUp_Flux.cpp>, line <231>, function <Flu_FixUp_Flux> !!
```
Upon checking the `Fluid/Flu_FixUp_Flux.cpp`, I noticed that the original [check](https://github.com/zengbs/gamer/blob/3307e30129128980d8da58000ae100471a32fa28/src/Fluid/Flu_FixUp_Flux.cpp#L230) used `SRHD_CheckUnphysical()`, while the [current approach](https://github.com/jzuhone/gamer/blob/89687e2dd6a13787780b976e36cdcf48ecd1486a/src/Fluid/Flu_FixUp_Flux.cpp#L231) employs `Hydro_CheckUnphysical()`.

The significant difference is that `SRHD_CheckUnphysical()` does NOT check the negative passive scalar (see [here](https://github.com/zengbs/gamer/blob/3307e30129128980d8da58000ae100471a32fa28/src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp#L609) in Tseng's original PR code and [here](https://github.com/jzuhone/gamer/blob/89687e2dd6a13787780b976e36cdcf48ecd1486a/src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp#L796) in your PR code). As a result, the behavior of `Fluid/Flu_FixUp_Flux.cpp` differs between the two code bases when there is a negative passive scalar.

As for which one is more accurate, perhaps @hyschive or @zengbs could provide their comments.

jzuhone commented:
@ChunYen-Chen Yes, I see them as well and I suspected this had something to do with it. But I am not certain why the scalars seem to advect correctly in @zengbs 's old version but not the new one.

hyschive commented:
@jzuhone @ChunYen-Chen I'll take a look at the issue of passive scalars.

ChunYen-Chen commented:
@hyschive
I suspect the issue might only occur when AMR is enabled. I tried the following scenarios
* `MAX_LEVEL = 2`:
   I got the error of
   ```
   ERROR: unphysical conserved variables at file <Fluid/Flu_FixUp_Flux.cpp>, line <231>, function <Flu_FixUp_Flux> !!
   ```
   in several time-steps. 
* `NX0_TOT_* = 256` and `MAX_LEVEL = 0`:
   There were no errors in the log file, at least up to 400 time-steps.
Hope the message is helpful for debugging.

hyschive commented:
@ChunYen-Chen Thanks for the information.

ChunYen-Chen commented:
Hi @jzuhone,
 I have submitted a pull request to update the Riemann test problem, enabling testing based on user-defined conditions. Please help review it at your earliest convenience. Thanks!

jzuhone commented:
@hyschive @ChunYen-Chen one other thing I should point out--the branch from @zengbs that I was using for the paper that worked with the passive scalars is https://github.com/zengbs/gamer/tree/cr-diffusion-x, which I am not sure is actually the branch used in either of his PRs.

hyschive commented:
@jzuhone Thanks for the additional information. I'll look into this issue asap (hopefully early next week).

jzuhone commented:
@hyschive @ChunYen-Chen it turns out that even though without refinement that the negative values for the passive scalars are gone, there are values that do not make much sense in the outputs, so the issue is probably independent of refinement.

jzuhone commented:
@hyschive @ChunYen-Chen I found the bug with the passive scalars and it is now fixed.

hyschive commented:
> @hyschive @ChunYen-Chen I found the bug with the passive scalars and it is now fixed.

@jzuhone Excellent! Thanks for your great effort and sorry for not being able to look into this issue in time.

jzuhone commented:
@hyschive @ChunYen-Chen AFAIK this is ready for another go-through, I believe I have addressed the open issues above. I am not sure what your plans are for further testing, let me know. There is at least nothing else I believe I need to add.

hyschive commented:
@jzuhone Thanks. I'll go through it again next week. Meanwhile, @JeiwHuang will help verify that this PR doesn't break any original non-SRHD tests. BTW, @ChunYen-Chen and @xuanweishan have been implementing a regression test for GAMER but, unfortunately, it's not ready yet.

JeiwHuang commented on include/TestProb.h:
```diff
@@ -40,8 +40,9 @@ extern double (*Mis_GetTimeStep_User_Ptr)( const int lv, const double dTime_dt )
 extern void (*Aux_Record_User_Ptr)();
-extern void (*BC_User_Ptr)( real fluid[], const double x, const double y, const double z, const double Time,
-                            const int lv, double AuxArray[] );
+extern void (*BC_User_Ptr)( real Array[], const int ArraySize[], real fluid[], const int NVar_Flu,
```
Probably because of this change, there is an error at `TestProblem/Hydro/Bondi/Init_TestProb_Hydro_Bondi.cpp(636)`: 
```
 error: a value of type "void (*)(real={float} *, double, double, double, dou
ble, int, double *)" cannot be assigned to an entity of type "void (*)(real={float} *, const int *, real={float} *, int, int, const int 
*, const double *, double, int, const int *, double *)"
     BC_User_Ptr              = SetGridIC;
                              ^

```
when compiling with `--model=HYDRO --fftw=FFTW3 --gravity=true`

JeiwHuang commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_DualEnergy.cpp:
```diff
@@ -153,7 +153,8 @@ real Hydro_Con2Dual( const real Dens, const real MomX, const real MomY, const re
 // --> note that DE_ENPY only works with EOS_GAMMA, which does not involve passive scalars
    Pres = Hydro_Con2Pres( Dens, MomX, MomY, MomZ, Engy, NULL, CheckMinPres_No, NULL_REAL, Emag,
-                          EoS_DensEint2Pres, EoS_AuxArray_Flt, EoS_AuxArray_Int, EoS_Table, NULL );
+                          EoS_DensEint2Pres, EoS_GuessHTilde, EoS_HTilde2Temp,
```
"EoS_GuessHTilde" and "EoS_HTilde2Temp" are undefined when compiling with `DUAL_ENERGY`  enabled (`--model=HYDRO --dual=DE_ENPY`)

JeiwHuang reviewed (commented):
Here are some compilation errors I encountered so far.

jzuhone commented:
@JeiwHuang thanks for pointing the compilation problems out, I believe they are resolved now. Please check.

JeiwHuang reviewed (approved):
@hyschive @jzuhone 
I have finished the tests on the original non-SRHD test problems.

- This PR can be compiled with various combinations of compilation options.

- This PR can reproduce the identical results as the main branch (except there is a bug in the analytical solution of temperature, I put it in the comment) for the following test problems:
`BlastWave`, `CDM_LSS`, `ClusterMerger`, `CR_Diffusion`, `JeansInstability`, `Jet`, `MHD_LinearWave`, `MHD_OrszagTangVortex`, `ParticleEquilibriumIC`, `Plummer`, `Riemann`, `SphericalCollapse`, `Zeldovich`.
- The bitwise reproducibility for test problems without SRHD is also verified for (i) with/without MPI and (ii) fresh run/restart.

jzuhone commented:
@JeiwHuang thanks for the further comments, I believe I have resolved them.

hyschive commented:
@JeiwHuang Thanks for reviewing it! Could you help finish the following two tasks?
- Confirm the bitwise reproducibility for the SRHD `JetICMWall` test problem (with/without MPI and fresh run/restart)
- File a feature request to add the SRHD Riemann problem

@jzuhone I'll take another look at the entire PR asap before merging.

JeiwHuang commented:
@hyschive 
I have confirmed the bitwise reproducibility for the SRHD `JetICMWall` test problem (with/without MPI and fresh run/restart).

jzuhone commented:
@hyschive should we try reproducing the Riemann problem, as suggested in #268, before merging?

hyschive commented:
> @hyschive should we try reproducing the Riemann problem, as suggested in #268, before merging?

@jzuhone: @JeiwHuang and @ChunYen-Chen have confirmed that the Riemann problem in the SRHD paper can be reproduced successfully by this PR. The only obstacle is that it requires outputting the SRHD primitive variables (rho, U_x/y/z, pressure) by hard-coding `Output_DumpData_Part.cpp` (see the 4th point in #268). So it would be better to add options similar `OPT__OUTPUT_POT` for SRHD-specific variables. Do you think you can do that in this PR (@JeiwHuang can deal with 1-3 points in #268)? If you are also busy, I would suggest dealing with that in a follow-up PR later as both JeiwHuang and ChunYen-Chen are unavailable recently.

jzuhone commented:
@hyschive I can add the output options to this PR tomorrow, yes.

jzuhone commented:
@hyschive I've checked the updates to `Output_DumpData_Part.cpp` and `Output_DumpData_Total_HDF5.cpp` against the `JetICMWall` problem and everything looks good now.

JeiwHuang commented:
@jzuhone
The new output options work well and I can use them to plot the SRHD Riemann problem. Thanks!

Just a minor issue:
I encountered this warning when compiling with `SUPPORT_HDF5`.
```
Output/Output_DumpData_Total_HDF5.cpp(1029): warning #823: reference is to variable "v" (declared at line 763) -- under old for-init scoping rules it would have been variable "v" (declared at line 1023)
  		     if ( v == LorentzDumpIdx )
  		          ^

Output/Output_DumpData_Total_HDF5.cpp(1032): warning #823: reference is to variable "v" (declared at line 763) -- under old for-init scoping rules it would have been variable "v" (declared at line 1023)
                       else if ( v >= VelDumpIdx0  &&  v < VelDumpIdx0+3 )
                                 ^

Output/Output_DumpData_Total_HDF5.cpp(1032): warning #823: reference is to variable "v" (declared at line 763) -- under old for-init scoping rules it would have been variable "v" (declared at line 1023)
                       else if ( v >= VelDumpIdx0  &&  v < VelDumpIdx0+3 )
                                                       ^
```

jzuhone commented:
@JeiwHuang good catch, fixed it.

jzuhone commented:
@hyschive anything else we should check/add here before merging?

hyschive commented:
@jzuhone I'm taking a final check and already going through half of the files. So far I only have some minor comments and will try to finish the review early next week (I need to write some recommendation letters and set the final exam paper first...).

hyschive reviewed (changes requested):
@jzuhone Thanks for your great effort in this PR! Sorry for my slow progress in reviewing it. I've added several comments, mostly minor. I list the major comments below.
- [x] HLLE doesn't support passive scalars.
- [x] The implementation of Eq. A1 in `CPU_Shared_FluUtility.cpp` is incorrect (the formula in the paper is correct but the implementation in this PR is wrong). @zengbs has confirmed that. The correct formula has been added in the comment. [EDITED]
- [x] Eq. A4 in the paper is incorrect. The correct formula has been added in the comment. [EDITED]
- [x] Why convert four-velocity to three-velocity in `CPU_dtSolver_HydroCFL.cpp` (lines 159-161)? @zengbs What do you think? 
- [ ] How about supporting `OPT__OUTPUT_ENTHANPY` to output the reduced enthalpy and enabling it by default? This will make computing primitive variables in post-processing much easier.
  - [ ] Relatedly, I'm not sure how it is currently done in yt but it would be great if we could update the gamer frontend to support computing primitive variables using the stored reduced enthalpy.
  - [ ] Relatedly, note that for SRHD the gas temperature is always stored (with `OPT__OUTPUT_TEMP`) in k_B*T/mc^2. It will be great if the gamer frontend in yt can support converting that to kelvin using the given mean molecular weight (`pc.mh`). 
- [x] Solve the merge conflicts.
- [x] [Minor] Remove trailing spaces and tabs introduced in this PR.

hyschive commented on src/Model_Hydro/CPU_Hydro/CPU_dtSolver_HydroCFL.cpp:
```diff
@@ -133,12 +155,24 @@ void CPU_dtSolver_HydroCFL  ( real g_dt_Array[], const real g_Flu_Array[][FLU_NI
          CFLy             = SQRT( CFLy );
          CFLz             = SQRT( CFLz );
+#        elif ( defined SRHD )
+         Pri[1] /= LorentzFactor;
+         Pri[2] /= LorentzFactor;
+         Pri[3] /= LorentzFactor;
```
@jzuhone : @zengbs has confirmed that these three lines are incorrect. Please remove them.

hyschive commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,336 @@
+      real C       = (real)43.0*M_Dsqr + (real)63.0;
+      real F       = (real)75.0*B;
+      real G       = (real)125.0*B*C*(*Constant);  
+      GuessHTilde  = SQRT(G);
```
I think this line should be replaced with `GuessHTilde  = G;`. @ChunYen-Chen Can you cross-check it?

hyschive commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,325 @@
+     real F       = (real)75.0*B;
+     real G       = (real)125.0*B*C*(*Constant);  
+     GuessHTilde  = SQRT(G);
+     GuessHTilde /= C*(F+SQRT(G+F*F));
```
I think this can be easily fixed by replacing `GuessHTilde  = SQRT(G);` with `GuessHTilde  = G;`. @ChunYen-Chen Can you cross-check it?

ChunYen-Chen replied:
Sorry about that I made a mistake in the last line of the `SRHD Solving A3`, so the solution of Eq. A3 is actually consistent with the original code written by Po-Hsun. Can you fix the `GuessHTilde` to previous version?

Here is the fixed write-up:
<img width="549" alt="Screenshot 2023-12-30 at 09 31 02" src="https://github.com/gamer-project/gamer/assets/70311975/4c3f2933-5945-498e-927b-3847f645af09">

hyschive replied:
@ChunYen-Chen Good. So we can simply replace `GuessHTilde  = SQRT(G);` with `GuessHTilde  = G;`, right?

ChunYen-Chen replied:
@hyschive Yes, we can fix it by replacing `GuessHTilde  = SQRT(G);` with `GuessHTilde  = G;`

hyschive commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:
```diff
@@ -383,6 +516,97 @@ void Hydro_Con2Flux( const int XYZ, real Flux[], const real In[], const real Min
+  if ( DiffFunc != NULL )
+
+  *DiffFunc = (real)2.0*H - (real)2.0*Temp - (real)2.0*H*DiffTemp +
+	        ( (real)2.0*Temp*DiffTemp*H*H - (real)2.0*Temp*Temp*H ) / SQR( Factor0 );
```
I thnk we should replace `( (real)2.0*Temp*DiffTemp*H*H - (real)2.0*Temp*Temp*H )` with `(real)2.0*Temp*H*( DiffTemp*H*Factor0 + Temp*MSqr_DSqr )`. @ChunYen-Chen Can you help cross-check it?

zengbs replied:
@jzuhone @ChunYen-Chen:

I have confirmed that @hyschive 's expression is correct.

Is it possible to replace the line

` ( (real)2.0*Temp*DiffTemp*H*H - (real)2.0*Temp*Temp*H ) / SQR( Factor0 );`

by

`(real)2.0*Temp*H*( DiffTemp*H*Factor0 + Temp*MSqr_DSqr ) / SQR( Factor0 );`.

Thanks!

hyschive commented:
@jzuhone Thanks for your great contribution to this PR. Any chance that we can merge it recently?

Note that all the incorrect equations have been addressed in the related comments.

jzuhone commented:
@hyschive sorry, got sidetracked by some other stuff. I think we can do it soon. 

I do have a question for you in one of your conversations about the passive scalars that needs resolving before we can finish.

hyschive commented:
@jzuhone Which conversation are you referring to?

jzuhone commented:
@hyschive original comment thread starts here: https://github.com/gamer-project/gamer/pull/185#discussion_r1434634522

hyschive commented:
> @hyschive original comment thread starts here: [#185 (comment)](https://github.com/gamer-project/gamer/pull/185#discussion_r1434634522)

@jzuhone Could you repeat your question? My comment therein is to make gas mass density and passive scalar density more consistent when converting between conserved and primitive variables.

jzuhone commented:
@hyschive in response to:

> Change this line as follows to improve performance since we also need _LorentzFactor when converting passive scalars.

```
const real _LorentzFactor = real(1.0) / LorentzFactor;
Out[0] *= _LorentzFactor;
```

I asked:

> I don't think that Out[0] is assigned a value at this line 238 yet, correct?

hyschive commented on src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp:
```diff
@@ -180,14 +215,43 @@ void Hydro_Con2Pri( const real In[], real Out[], const real MinPres,
+
+   if ( LorentzFactorPtr != NULL )   *LorentzFactorPtr = LorentzFactor;
+
+   Out[0] = In[0]/LorentzFactor;
```
> I don't think that Out[0] is assigned a value at this line 238 yet, correct?

You are absolutely correct. Sorry for my careless comment! Let's change it to `Out[0] = In[0] * _LorentzFactor;`.

BTW, somehow I cannot see the reply " I don't think that Out[0] is assigned a value at this line 238 yet, correct?". Not sure if it's caused by a bug on GitHub.

jzuhone commented:
@hyschive will the gamma-law EOS ever be implemented for SRHD? If not, we can simplify the yt code considerably.

hyschive commented:
> @hyschive will the gamma-law EOS ever be implemented for SRHD? If not, we can simplify the yt code considerably.

I don't think we have a strong motivation to implement the gamma-law EoS recently.

jzuhone commented:
@hyschive see my answer to your comment here, it's the last one for now: https://github.com/gamer-project/gamer/pull/185/files#r1434657969

hyschive commented:
> @hyschive see my answer to your comment here, it's the last one for now: https://github.com/gamer-project/gamer/pull/185/files#r1434657969

@jzuhone Somehow I still cannot see your answer. Could you repost it here?

jzuhone commented:
@hyschive Here it is. You said: 

"Similarly, move these passive-scalar operations to below line 404 (#  else) to avoid introducing too many #ifdef SRHD checks. I think it would help improve readability."

I said in response:

"what about the note "do it before invoking EoS_DensPres2Eint() since the latter requires the mass density instead of mass fraction of passive scalars"

It seems as if we cannot move these lines below line 404 for this very reason."

hyschive commented:
@jzuhone I guess line 404 has changed. What I mean is to move

```
// passive scalars
// --> do it before invoking EoS_DensPres2Eint() since the latter requires the mass density
//     instead of mass fraction of passive scalars
#  if ( NCOMP_PASSIVE > 0 )
// copy all passive scalars
   for (int v=NCOMP_FLUID; v<NCOMP_TOTAL; v++)  Out[v] = In[v];

// convert the mass fraction of target passive scalars back to mass density
   if ( FracPassive )
      for (int v=0; v<NFrac; v++)   Out[ NCOMP_FLUID + FracIdx[v] ] *= In[0];
#  endif
```

to the very beginning of the following code section (so before calling `EoS_DensPres2Eint()`).

``` 
#  else

** PUT HERE **

// primitive --> conserved
   Out[0] = In[0];
   Out[1] = In[0]*In[1];
   Out[2] = In[0]*In[2];
   Out[3] = In[0]*In[3];

...

#  endif
```

jzuhone commented:
Hi @hyschive -- I believe I have resolved everything. Please especially have a look at commits https://github.com/gamer-project/gamer/pull/185/commits/8dffa5956f9859175dc87f8bc49a785de8a4f204 through https://github.com/gamer-project/gamer/pull/185/commits/56af5fb805790e03cf6301ae87d5d126a2760621. 

One note--I have decided that for consistency with non-SR hydro we should store `Temp` in Kelvin, so I have left the conversion in `Hydro_Con2Temp` in but removed it elsewhere. This will require an update to yt, which I have already done but not submitted the PR. 

I believe that this is good to go.

hyschive commented:
@jzuhone Thanks! I'll start to review it.

hyschive commented on src/EoS/TaubMathews/CPU_EoS_TaubMathews.cpp:
```diff
@@ -0,0 +1,336 @@
+#ifndef __CUDACC__
+void EoS_SetAuxArray_TaubMathews( double AuxArray_Flt[], int AuxArray_Int[] )
+{
+   AuxArray_Flt[0] = ( OPT__UNIT ) ? MOLECULAR_WEIGHT * Const_amu / Const_kB * (UNIT_E/UNIT_M)
```
Replace `Const_amu` with `MU_NORM`, which has a default value of `Const_mH` instead of `Const_amu` to be more consistent with the gamer frontend in yt (though the values of `Const_mH` in gamer and yt are still not fully consistent).

hyschive commented on src/TestProblem/Hydro/AcousticWave/Init_TestProb_Hydro_AcousticWave.cpp:
```diff
@@ -115,24 +125,40 @@ void SetParameter()
 // (2) set the problem-specific derived parameters
+
+// must initialize EoS first
+   EoS_Init();
```
Enclose it by `#ifdef SRHD ... #endif`

hyschive commented on example/test_problem/Template/Input__Parameter:
```diff
@@ -98,6 +98,7 @@ HUBBLE0                       0.70        # dimensionless Hubble parameter (curr
 DT__MAX                      -1.0         # dt criterion: maximum allowed dt (<0=off) [-1.0]
 DT__FLUID                    -1.0         # dt criterion: fluid solver CFL factor (<0=auto) [-1.0]
 DT__FLUID_INIT               -1.0         # dt criterion: DT__FLUID at the first step (<0=auto) [-1.0]
+DT__SPEED_OF_LIGHT              0         # compute time step based on speed of light ## SRHD ONLY ##
```
Add `[0]` before `##SRHD ONLY##` to clarify the default value.

hyschive reviewed (changes requested):
@jzuhone I've added some final comments. Important ones are the compilation issues of the MHM scheme.

jzuhone commented:
@hyschive in case you didn't see it, here's my reply to your comment about mH vs. amu: 

You said:

"Replace Const_amu with MU_NORM, which has a default value of Const_mH instead of Const_amu to be more consistent with the gamer frontend in yt (though the values of Const_mH in gamer and yt are still not fully consistent)."

I replied:
"I was actually planning on changing it in yt to amu"

hyschive commented:
> @hyschive in case you didn't see it, here's my reply to your comment about mH vs. amu:
> 
> You said:
> 
> "Replace Const_amu with MU_NORM, which has a default value of Const_mH instead of Const_amu to be more consistent with the gamer frontend in yt (though the values of Const_mH in gamer and yt are still not fully consistent)."
> 
> I replied: "I was actually planning on changing it in yt to amu"

@jzuhone I think we should still replace `Const_amu` with `MU_NORM` to be self-consistent. Once the normalization in yt is changed to amu, we can change the default value of `MU_NORM` accordingly. What do you think? 

BTW, thanks for reposting the reply here. I still cannot see your inline comments.

jzuhone commented:
> BTW, thanks for reposting the reply here. I still cannot see your inline comments.

@hyschive silly question, but have you tried refreshing the browser window? Very weird.

jzuhone commented:
@hyschive do you have a preference for amu or mH in the definition?

hyschive commented:
@jzuhone 
> silly question, but have you tried refreshing the browser window? Very weird.

Doesn't help. And @JeiwHuang cannot see your reply either.

> do you have a preference for amu or mH in the definition?

I don't have a preference. You make the call and I'll modify the default value of `MU_NORM` accordingly.

jzuhone commented:
@hyschive I've addressed all of these comments. The code compiles correctly for both SRHD and HD. 

One thing you should probably definitely check in detail--in `Aux_CheckConservation` I used a calculation for the kinetic energy for SRHD specified [here](https://github.com/yt-project/yt/files/5871557/gamer.pdf) by @zengbs. It's the same calculation that the yt GAMER frontend uses. The drawback is that it calls both `Hydro_Con2Pri` and `Hydro_Con2HTilde` and thus may be expensive. Up to you if you think it's worth it. 

I've got an updated version of the GAMER frontend in yt based on these changes which appears to work--before I make a PR to the yt repo I'm going to test it a bit more.

hyschive commented on src/Auxiliary/Aux_Check_Conservation.cpp:
```diff
@@ -176,13 +181,36 @@ void Aux_Check_Conservation( const char *comment )
+                                 EoS.DensPres2Eint_FuncPtr, EoS.GuessHTilde_FuncPtr, EoS.HTilde2Temp_FuncPtr, 
+                                 EoS.AuxArrayDevPtr_Flt, EoS.AuxArrayDevPtr_Int, EoS.Table, NULL, &Lrtz );
+
+                  HTilde       = Hydro_Con2HTilde( Cons, EoS.GuessHTilde_FuncPtr, EoS.HTilde2Temp_FuncPtr, 
+                                                   EoS.AuxArrayDevPtr_Flt, EoS.AuxArrayDevPtr_Int, EoS.Table );   
```
[Bug] Similarly, replace `EoS.GuessHTilde_FuncPtr, EoS.HTilde2Temp_FuncPtr, EoS.AuxArrayDevPtr_Flt, EoS.AuxArrayDevPtr_Int, EoS.Table` with `EoS_GuessHTilde_CPUPtr, EoS_HTilde2Temp_CPUPtr, EoS_AuxArray_Flt, EoS_AuxArray_Int, h_EoS_Table`.

hyschive reviewed (changes requested):
@zengbs Following the discussion [here](https://github.com/gamer-project/gamer/pull/185#issuecomment-1891297791), I have a few quick questions about your [write-up](https://github.com/yt-project/yt/files/5871557/gamer.pdf) of the SRHD formulae.
- Thermal energy density `E_thermal` in Eq. [9] is defined at the local rest frame, right?
- Kinetic energy density in Eq. [12] is however defined in the lab frame and is equivalent to `E - Dc^2 - \gamma*E_thermal`, where `E` is the total energy density in Eq. 11 and `\gamma` is the Lorentz factor. Is that right?
- Should we record `\gamma*E_thermal` or just `E_thermal` in `Aux_Check_Conservation.cpp`?

@jzuhone
- There are some bugs in the EoS functions in `Aux_Check_Conservation.cpp`.
- Let's enable `OPT__OUTPUT_TEMP = 1` by default for SRHD.

jzuhone commented:
@hyschive it seems as if given these complications we should not try to record Ekin, Eint in `Aux_CheckConservation`. What do you think?

hyschive commented:
@jzuhone The fixes should be straightforward but I'm not one hundred percent sure about these SRHD formulae off-hand. If we do not receive confirmation from @zengbs in the next 24 hours, I suggest commenting out (but not deleting) the lines of computing `Ekin` and `Eint` and fixing both values to zero for now. I'll then add that to the SRHD to-do list.

jzuhone commented:
@hyschive shall I comment out those lines?

hyschive commented:
@jzuhone Yes

hyschive commented:
@jzuhone I'll merge this PR once those lines in `Aux_Check_Conservation.cpp` are commented out and replaced with `Ekin = Eint = 0.0`. Meanwhile, @JeiwHuang is helping perform some final checks to make sure that those corrected SRHD equations do not lead to any unexpected results.

jzuhone commented:
@hyschive done

JeiwHuang reviewed (commented):
I have completed some tests for this PR, and here are my results:

- This branch can be compiled without errors using different combinations of the compilation options of HYDRO and SRHD.
- This branch can reproduce the same results as the main branch for the non-SRHD test problems (`BlastWave`, `CDM_LSS`, `ClusterMerger`, `CR_Diffusion`, `JeansInstability`, `Jet`, `MHD_LinearWave`, `MHD_OrszagTangVortex`, `ParticleEquilibriumIC`, `Plummer`, `Riemann`, `SphericalCollapse`, `Zeldovich`), except for one bug related to the analytical solution of temperature in Output_L1Error.cpp.
- The results of the SRHD `AcousticWave` test problem show a similar L1 error scaling as my previous test.
- The results of the SRHD `Riemann` test problem in the ultra-relativistic limit are similar to my previous test and can match the analytical solution.
- The SRHD `Riemann` test problem in mixed limits is still running.
- I can run the SRHD `JetICMWall` test problem; however, the results are not the same as in my previous test. I have included details about this issue in the inline comment.

I have also placed some bugs and minor issues I found during the test in the inline comments.

hyschive commented on example/test_problem/Hydro/JetICMWall/yt_script/make_slice.py:
@jzuhone This is strange. It seems that either the Lorentz factor stored in a gamer snapshot is wrong or yt fails to compute the Lorentz factor. But I thought the only thing that changed recently was the unit of the output temperature. What do you think?

jzuhone commented:
@JeiwHuang @hyschive you need to use the code in this draft PR for yt https://github.com/yt-project/yt/pull/4782, which has been written to work with this GAMER PR.

hyschive commented:
@jzuhone OK. But just out of curiosity, do you have any idea what might cause this issue? I thought the only changes we applied recently were to (i) change the output temperature unit and (ii) output reduced enthalpy. Why would that break the gamer frontend in yt?

jzuhone commented:
@hyschive I'll have to investigate, but I suspect that there may be an issue here with how yt calculates the enthalpy and everything else from kT/mc^2 in the `main` branch, but now is receiving a number in Kelvin.

hyschive commented:
@jzuhone OK. I think I'll merge it anyway once the bug in `Outout_L1Error.cpp` is fixed.

hyschive reviewed (approved):
@jzuhone Looks good. I'll merge it right away. But it would be great if we could still figure out what breaks the current gamer frontend in yt (even though the new yt PR solves the problem).

hyschive commented:
@zengbs @jzuhone @ChunYen-Chen @JeiwHuang Thank you all for implementing/updating/testing/reviewing this SRHD PR. Let's merge it!

After that, I'll probably polish the code format slightly and then file a feature request to record the future SRHD work.

hyschive commented:
@jzuhone 
> as for the yt issue, I would maybe try making a copy of the HDF5 file, with the `Temp` field multiplied by k/mc^2, and then trying to compare to the previous version using the `main` yt branch? otherwise, we can just use the branch I linked earlier.

I believe I’ve fixed this issue [here](https://github.com/gamer-project/gamer/commit/fdaa8085995a1c9400c8da611d5d2da275b25d60).

hyschive commented:
I've updated wiki to include SRHD.
